Update IO data

MoLi7 opened the pull request:
This PR will include 
1. modified code and functions for downloading and saving IO data up to 2020 (most recent year BEA has provided)
2. updated `rda` of data and `json` of metadata

MoLi7 commented on inst/extdata/metadata/Detail_Make_2012_BeforeRedef_metadata.json:
```diff
@@ -1,13 +1,15 @@
   "ext": ["json"],
-  "date_created": ["2021-07-30 12:52:37"],
-  "tool_meta": {
+  "date_created": ["2022-03-28"],
```
this is when the `rda` file was created by `useeior`

MoLi7 commented on inst/extdata/metadata/Detail_Make_2012_BeforeRedef_metadata.json:
```diff
@@ -1,13 +1,15 @@
     "author": ["US Bureau of Economic Analysis"],
-    "source_url": ["https://edap-ord-data-commons.s3.amazonaws.com/useeior/AllTablesIO.zip"],
+    "source_url": ["https://apps.bea.gov/industry/iTables%20Static%20Files/AllTablesIO.zip"],
+    "date_last_modified": ["2021-09-27 08:51:30"],
```
this is when the original data was modified by provider, e.g. BEA

MoLi7 commented on inst/extdata/metadata/Detail_Make_2012_BeforeRedef_metadata.json:
```diff
@@ -1,13 +1,15 @@
-    "source_url": ["https://edap-ord-data-commons.s3.amazonaws.com/useeior/AllTablesIO.zip"],
+    "source_url": ["https://apps.bea.gov/industry/iTables%20Static%20Files/AllTablesIO.zip"],
+    "date_last_modified": ["2021-09-27 08:51:30"],
+    "date_obtained": ["2022-03-28"],
```
this is when the original data was obtained by `useeior`

WesIngwersen replied:
> @WesIngwersen could you review the sample metadata I pushed and my comments about the `date` fields? Thanks!

OK just made one change. thanks

MoLi7 commented on inst/extdata/metadata/Detail_Make_2012_BeforeRedef_metadata.json:
```diff
@@ -1,13 +1,15 @@
-  "date_created": ["2021-07-30 12:52:37"],
-  "tool_meta": {
+  "date_created": ["2022-03-28"],
+  "data_meta": {
```
I think `data_meta` is more appropriate than `tool_meta` for describing the content it contains.

WesIngwersen replied:
yes i think the `tool_meta`  is more relevant when the data are a product of the tool. that is why its used in flow-by datasets from flowsa

MoLi7 commented:
@WesIngwersen could you review the sample metadata I pushed and my comments about the `date` fields? Thanks!

MoLi7 reviewed (commented):
[8db23ee](https://github.com/USEPA/useeior/pull/221/commits/8db23ee6c666baa4b2691c4db259c11430ee6fd0) updates 2010-2020 Summary Make and Use

MoLi7 commented on inst/extdata/metadata/Detail_Import_2012_BeforeRedef_metadata.json:
```diff
@@ -1,13 +1,14 @@
     "author": ["US Bureau of Economic Analysis"],
     "source_url": ["https://apps.bea.gov/industry/xls/io-annual/ImportMatrices_Before_Redefinitions_DET_2007_2012.xlsx"],
-    "bib_id": [""]
+    "date_last_modified": ["unknown"],
```
Import matrix was accessed directly from a url on [this page](https://www.bea.gov/industry/input-output-accounts-data). There is no information about when the data was last modified.

WesIngwersen replied:
The text under the main header of that page states
"BEA’s use tables were updated on Feb. 11, 2022, to correct a small number of cells in the 2014-2020 use tables initially published on Sep 30, 2021. BEA’s requirements tables, supply tables, and PCE/PEQ bridge tables were also updated. This update does not affect the GDP by industry statistics."

I would assume the import matrices, since they are versions of the Use table, where updated also n Feb 11

MoLi7 replied:
The year range 2014-2020 seems to me that only **Summary** Use and Import matrix was updated on Feb 11.

MoLi7 commented:
FYI, preliminary test shows the updated BEA IO data do NOT affected USEEIO model results. @WesIngwersen 
- In the `develop` branch, build a `USEEIOv2.0.1-411` as `model_old`
- In the `updateIO` branch, build a `USEEIOv2.0.1-411` as `model_new`
- Compare model matrices including `B`, `D`, `M`, and `N` between `model_old` and `model_new` - differences == 0

WesIngwersen commented:
2021 fourth quarter GDP and associated industry account stats were just released, on [schedule](https://www.bea.gov/data/industries/gross-output-by-industry).  The interactive data explorer now has the 2021 GO, chain-type price index data, at least at a more aggregate level

WesIngwersen commented:
> FYI, `BEAData.R` and all BEA IO data have been updated completely. Preliminary test shows the updated BEA IO data do NOT affected USEEIO model results. @WesIngwersen
> 
> * In the `develop` branch, build a `USEEIOv2.0.1-411` as `model_old`
> * In the `updateIO` branch, build a `USEEIOv2.0.1-411` as `model_new`
> * Compare model matrices including `B`, `D`, `M`, and `N` between `model_old` and `model_new` - differences == 0

Thank you. I think we need to compare the summary level models since post-2012 data may have been updated. 
BEA stated that 
"BEA’s use tables were updated on Feb. 11, 2022, to correct a small number of cells in the 2014-2020 use tables initially published on Sep 30, 2021"
so the 2014-2018 Use tables at the summary level have likely changed from what we had stored in rda in useeior v1.0 releases

WesIngwersen commented:
My comparison of a model summary level commodity-based A matrix built with the old and new IO tables showed a difference in 87% of the cells. I think this deserves more exploration

MoLi7 commented:
> My comparison of a model summary level commodity-based A matrix built with the old and new IO tables showed a difference in 87% of the cells. I think this deserves more exploration

Thanks for catching the error. Did you manage to install useeior with the new data from remote?

MoLi7 commented:
FYI @WesIngwersen I figured that the error occurred in consecutive installation two versions of `useeior` is related to R session - one R session cannot handle install-uninstall-install of two (versions of) packages. To successfully install and load each version of `useeior`, you must start with a new R session (use `Ctrl` + `Shift` + `0`).

WesIngwersen commented:
Ok I see. I guess thats a potential plus for using Renv to create virtual environments - maybe each instance of a Renv has its own session?

MoLi7 commented:
I assume the workflow would be

1. solve conflicts
2. close this PR to bring updated IO data to `develop`
3. merge `develop` with `master`
4. publish the release `1.0.3` from master

But this release would include more than the updated IO data, because some disagg functions were modified, integrated hybrid models were added,`loadIOData` function is also modified to take in state IO data (partially). These may be changes unwanted by this particular release unless we treat it as a big dump of updates. @WesIngwersen

WesIngwersen commented:
@MoLi7 @bl-young  I think we might want to rather consider the alternative git model of pulling this branch into a release branch, then into master, rather than going through the develop branch

MoLi7 commented:
Yes we can go that route, and I suppose the new `release` branch would be created based off of `master`? However, this probably won't change the fact that `updateIO` has incorporated many other changes.

MoLi7 commented:
@WesIngwersen per our discussion, next steps for preparing the release are as follows.

1. Ciose https://github.com/USEPA/useeior/pull/216
2. Merge `develop` to `updateIO` to eliminate R-CMD-check warnings due to stateior package
3. Open a PR to merge `updateIO` to `master`, this will require resolution of a few conflicts.
4. Release a new version from `master`.

This new release will include many updates:

1. Updated national IO data
2. Added functionality to load state IO data and updated related functions
3. Added feature of integrated hybrid models
4. Maybe more...

Tagging @bl-young @jvendries for your reference and potential feedback.

bl-young commented:
This sounds fine to me. I believe there were a few minor fixes to support disaggregation of industry models (currently on develop). Also I will leave it up to you if we should pull #231 in at this time as well. (I don't understand why the actions are failing on that one). I have a few small commits to wrap up hybridization. (see [hybridization](https://github.com/USEPA/useeior/compare/develop...hybridization)). Please let me know which branch and when I should target that PR.

I also might flag for @jvendries that there is some commented out code in disaggregation.R that should be reviewed and removed most likely.

MoLi7 commented:
Thanks @bl-young. I'd prefer leaving #231 for next release as it's a fairly user-facing update that deserve some dedicated highlight. Thinking it can wait till #228 is resolved because they are similar enhancement, then the new features can be packed in a release. The actions are failing most probably due to `printValidationResults` is [called as an exported function](https://github.com/USEPA/useeior/blob/master/.github/workflows/R-CMD-check.yaml#L75) in R-CMD-check.yaml, but it's not exported by definition. I can export it to solve this problem.

Pulling `hybridization` to `develop` sounds good to me, see #232 .

I will wait to hear from @jvendries before merging any PRs.

bl-young commented:
> The actions are failing most probably due to `printValidationResults` is [called as an exported function](https://github.com/USEPA/useeior/blob/master/.github/workflows/R-CMD-check.yaml#L75) in R-CMD-check.yaml, but it's not exported by definition. I can export it to solve this problem.

I could have sworn that wasn't always the case. I thought in the past we have seen these results get printed out. Either way, that all sounds good.

jvendries commented:
> I will wait to hear from @jvendries before merging any PRs.

I think the changes to disaggregationFunctions.R are limited to removing commented code, as Ben mentioned, so it should be fine to merge this pull request.

MoLi7 commented:
Now `updateIO` is up-to-date with `develop`. To confirm it, see [here](https://github.com/USEPA/useeior/compare/updateIO...develop) and [here](https://github.com/USEPA/useeior/network).

An R-CMD-Check [Actions](https://github.com/USEPA/useeior/actions/runs/2721110590) is triggered. Will follow up with results later.

@WesIngwersen @bl-young @jvendries

MoLi7 commented:
Reminded by [this error](https://github.com/USEPA/useeior/runs/7475412007?check_suite_focus=true#step:6:76) that we need to add `tibble` under _Imports_ in DESCRIPTION. Suppose @a-w-beck had 3.1.6 (or 3.1.7) when [deploying it](https://github.com/USEPA/useeior/commit/ec31e7b8cedb0111bb646d5b80743baf66101a15) in May?

bl-young commented:
@MoLi7 looks like there are some new validation issues on all models, but especially in the disaggregated models. I expect that with the data update to the BEA tables that the disaggregation input data will need to be updated.

MoLi7 commented:
> @MoLi7 looks like there are some new validation issues on all models, but especially in the disaggregated models. I expect that with the data update to the BEA tables that the disaggregation input data will need to be updated.
> 
> Could run the action on develop first to confirm.

Actually there is no validation issues in the [CMD-check](https://github.com/USEPA/useeior/actions/runs/2721110590) based off of 61f6256.

I ran another check on `develop` (see [here](https://github.com/USEPA/useeior/actions/runs/2721266179)) and got the error about `printValidationResults` not being exported. But it's already fixed in 87bbe70 and has not returned the same error after that.

Let me fix the `tibble` issue and run CMD-check one more time.

MoLi7 commented:
Everything looks good in this latest [R-CMD-check](https://github.com/USEPA/useeior/actions/runs/2721352764).

bl-young commented:
some of the sectors are failing the validations: https://github.com/USEPA/useeior/runs/7476060881?check_suite_focus=true#step:8:100

MoLi7 commented:
> some of the sectors are failing the validations: https://github.com/USEPA/useeior/runs/7476060881?check_suite_focus=true#step:8:100

Ahhhh that's right. I wasn't looking into the results. Thanks for catching them. Could you or Jorge address them?

bl-young commented:
> Ahhhh that's right. I wasn't looking into the results. Thanks for catching them. Could you or Jorge address them?

Yes we can look into it. Though there are some failures in non-disaggregated models as well.

MoLi7 commented:
> Yes we can look into it. Though there are some failures in non-disaggregated models as well.

Thanks @bl-young ! I'm looking into the non-disagg models now.

MoLi7 commented on inst/extdata/metadata/Detail_GrossOutput_IO_metadata.json:
```diff
@@ -1,9 +1,9 @@
   "tool": ["useeior"],
   "name_data": ["Detail_GrossOutput_IO"],
-  "tool_version": ["1.0.1"],
+  "tool_version": ["1.0.3"],
```
This data and metadata was updated to address a sector mismatch caused by invalid values in mapping table.
But the more important takeaway is that updated IO data were created under `useeior` 1.0.1. I think we should change all of them to 1.0.3 to be consistent with this release. @WesIngwersen

bl-young replied:
I wonder if this release would be better suited as `1.1` given the update to the BEA data? We expect this to change the results some right for existing models?

MoLi7 replied:
Yes, it makes sense to use `1.1`, thanks @bl-young !

@WesIngwersen I realized that we didn't have a GH actions workflow for downloading and saving BEA data, and that's why `date_accessed` wasn't changed. I will automate it with a new actions workflow.

MoLi7 commented:
> Yes we can look into it. Though there are some failures in non-disaggregated models as well.

There was a sector mismatch that caused no industry output value for sector 424200 and later led to that validation failure. Now it's been fixed and may have also addressed a few failures in the first disagg model validation (see [here](https://github.com/USEPA/useeior/runs/7476060622?check_suite_focus=true)).

@bl-young here is still one failure with 562HAZ/US which I will leave to you. :)

MoLi7 commented:
@bl-young [7939d97](https://github.com/USEPA/useeior/pull/221/commits/7939d977bde67f0b1c3f48290e970b5d6e8197d7) brought the changes you made in #234 to `develop` then to `updateIO`.

I know there is a remaining issue #233 and @jvendries can work in `develop` or another branch to fix it, but I want to modify this PR to target `master` instead of `develop` in order to enable automatic CI/CD runs.

a-w-beck commented:
> Reminded by [this error](https://github.com/USEPA/useeior/runs/7475412007?check_suite_focus=true#step:6:76) that we need to add `tibble` under _Imports_ in DESCRIPTION. Suppose @a-w-beck had 3.1.6 (or 3.1.7) when [deploying it](https://github.com/USEPA/useeior/commit/ec31e7b8cedb0111bb646d5b80743baf66101a15) in May?

@MoLi7 I believe I had 3.1.5, but given the current `dplyr >= 1.0.6` requirement I'd recommend just `tibble >= 2.1.3` (retreived from dplyr 1.0.6's DESCRIPTION file)

MoLi7 commented:
Thanks everyone for your inputs. This PR is ready to be merged.
